File: frontend/download_models.py
#!/usr/bin/env python3
"""
Download face-api.js model files to local storage
Uses multiple CDN sources for reliability
"""

import os
import urllib.request
import urllib.error
import ssl

def download_models():
    models_dir = os.path.join("public", "models")
    os.makedirs(models_dir, exist_ok=True)
    
    print("\n" + "="*70)
    print("📥 DOWNLOADING FACE-API MODELS")
    print("="*70 + "\n")
    
    # Model files we need
    files_to_download = [
        "tiny_face_detector_model-weights_manifest.json",
        "tiny_face_detector_model-weights.weights.bin",
        "face_expression_model-weights_manifest.json",
        "face_expression_model-weights.weights.bin",
    ]
    
    # CDN URLs to try (in order of preference)
    cdn_urls = [
        "https://unpkg.com/face-api.js@0.22.2/dist/models",
        "https://cdn.jsdelivr.net/npm/face-api.js@0.22.2/dist/models",
    ]
    
    ssl_context = ssl.create_default_context()
    ssl_context.check_hostname = False
    ssl_context.verify_mode = ssl.CERT_NONE
    
    downloaded_count = 0
    
    for filename in files_to_download:
        filepath = os.path.join(models_dir, filename)
        
        # Skip if already exists
        if os.path.exists(filepath) and os.path.getsize(filepath) > 0:
            size = os.path.getsize(filepath)
            print(f"✓ {filename} ({size:,} bytes)")
            downloaded_count += 1
            continue
        
        print(f"↓ {filename}...", end=" ", flush=True)
        
        downloaded = False
        for base_url in cdn_urls:
            if downloaded:
                break
                
            url = f"{base_url}/{filename}"
            try:
                req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
                with urllib.request.urlopen(req, context=ssl_context, timeout=15) as response:
                    content = response.read()
                    if len(content) > 100:  # Models should be > 100 bytes
                        with open(filepath, 'wb') as f:
                            f.write(content)
                        size = len(content)
                        print(f"✓ ({size:,} bytes)")
                        downloaded = True
                        downloaded_count += 1
            except (urllib.error.HTTPError, urllib.error.URLError, OSError):
                continue
        
        if not downloaded:
            print("✗ Not found on CDN")
    
    print("\n" + "="*70)
    print(f"📊 Downloaded: {downloaded_count}/{len(files_to_download)} files")
    print("="*70)
    
    if downloaded_count > 0:
        print(f"\n✅ Models saved to: public/models/")
    else:
        print("\n⚠️  Models not found on CDN")
        print("App will use CDN at runtime with automatic fallback\n")
    
    return downloaded_count > 0

if __name__ == "__main__":
    try:
        download_models()
    except Exception as e:
        print(f"\n❌ Error: {e}")

File: frontend/download-models.js
/**
 * Download face-api.js models locally
 * Run: node download-models.js
 * This will download all required model files to ./public/models/
 */

import fs from 'fs';
import path from 'path';
import https from 'https';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const MODELS_DIR = path.join(__dirname, 'public', 'models');

// Ensure directory exists
if (!fs.existsSync(MODELS_DIR)) {
  fs.mkdirSync(MODELS_DIR, { recursive: true });
  console.log(`✅ Created directory: ${MODELS_DIR}`);
}

// List of model files to download
const MODEL_FILES = [
  'tiny_face_detector_model-weights_manifest.json',
  'tiny_face_detector_model-weights.weights.bin',
  'face_expression_model-weights_manifest.json',
  'face_expression_model-weights.weights.bin',
];

const BASE_URL = 'https://cdn.jsdelivr.net/npm/face-api.js@0.22.2/dist/models';

/**
 * Download a file from URL
 */
function downloadFile(url, filePath) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(filePath);
    
    https.get(url, (response) => {
      if (response.statusCode !== 200) {
        reject(new Error(`Failed to download ${url}: ${response.statusCode}`));
        return;
      }

      response.pipe(file);

      file.on('finish', () => {
        file.close();
        console.log(`✅ Downloaded: ${path.basename(filePath)}`);
        resolve();
      });
    }).on('error', (err) => {
      fs.unlink(filePath, () => {}); // Delete incomplete file
      reject(err);
    });

    file.on('error', (err) => {
      fs.unlink(filePath, () => {}); // Delete incomplete file
      reject(err);
    });
  });
}

/**
 * Download all models
 */
async function downloadAllModels() {
  console.log('\n═══════════════════════════════════════════════════════════');
  console.log('📥 DOWNLOADING FACE-API MODELS');
  console.log('═══════════════════════════════════════════════════════════\n');

  try {
    for (const file of MODEL_FILES) {
      const url = `${BASE_URL}/${file}`;
      const filePath = path.join(MODELS_DIR, file);

      // Skip if already exists
      if (fs.existsSync(filePath)) {
        console.log(`⏭️  Already exists: ${file}`);
        continue;
      }

      console.log(`📥 Downloading: ${file}...`);
      await downloadFile(url, filePath);
    }

    console.log('\n═══════════════════════════════════════════════════════════');
    console.log('✅ ALL MODELS DOWNLOADED SUCCESSFULLY!');
    console.log('═══════════════════════════════════════════════════════════\n');
    console.log('📁 Models location: ./public/models/');
    console.log('\nModels are now ready for local use.');
    console.log('The app will use: /models/ (local) instead of CDN\n');
  } catch (error) {
    console.error('\n❌ ERROR downloading models:');
    console.error(error.message);
    process.exit(1);
  }
}

// Run download
downloadAllModels();
